enable mkl_gemm_bf16bf16f32 in cpublas::gemm

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105131
* #107128
* #104863
* #104693
* #103826
* #104584
* __->__ #104583

This one is a wrapper upon `mkl_gemm_bf16bf16f32` which is used in flash attention kernel on intel 4th gen xeon.
Fallback path has also been implemented on cpublas::gemm in case `mkl_gemm_bf16bf16f32` is not available.

ref: https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-0/cblas-gemm-bf16bf16f32.html

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

mingfeima committed:
enable mkl_gemm_bf16bf16f32 in cpublas::gemm

[ghstack-poisoned]

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




[ghstack-poisoned]

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




[ghstack-poisoned]

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




[ghstack-poisoned]

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




[ghstack-poisoned]

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




[ghstack-poisoned]

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




[ghstack-poisoned]

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




[ghstack-poisoned]

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




[ghstack-poisoned]

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




[ghstack-poisoned]

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/CPUBlas.cpp:
```diff
@@ -327,6 +327,36 @@ void gemm(
+#ifdef MKL_HAS_SBGEMM
+  if (use_blas_gemm(transa, transb, m, n, k, lda, ldb, ldc)) {
+    int m_ = m, n_ = n, k_ = k, lda_ = lda, ldb_ = ldb, ldc_ = ldc;
+    mkl_gemm_bf16bf16f32(transa, transb, m_, n_, k_, alpha, a, lda_, b, ldb_, beta, c, ldc_);
```
Can we just use `sbgemm_` directly in `CPUBlas.cpp`, which is already used in the all-`BFloat16` overload?

mingfeima replied:
No, I can not directly use `sbgemm_` in `CPUBlas.cpp`, since `sbgemm_` is from OpenBlas and MKL doesn't have a routine with the same name.

But I just added a path with `sbgemm_` in this function as well, so that if PyTorch is built against OpenBlas, it will also receive speedup using the scaled dot product attention optimization (e.g. flash attention).

peterbell10 commented on aten/src/ATen/native/CPUBlas.cpp:
```diff
@@ -327,6 +327,36 @@ void gemm(
+  }
+#endif
+  int64_t c_size = n * ldc;
+  std::vector<at::BFloat16> bfloat_v(c, c + c_size);
```
This truncates the bias unnecessarily. Could we not do `gemm` with `beta=0`, then add it to the bias in full float precision?

mingfeima replied:
Updated, first do the gemm with beta == 0 and then update C in float32.

jgong5 replied:
Have you updated the code? I am still seeing you pass the `beta`? Also, with `beta == 0`, you don't have to initialize the `bfloat_v`.

peterbell10 replied:
Looks like the commit from @Valentine233 undid the changes

mingfeima replied:
yes, the previous change has been undid by @Valentine233, her kernels of flash attention depends on this feature.

I have created a new one https://github.com/pytorch/pytorch/pull/107196 , could you please review again ? thanks!

peterbell10 commented on aten/src/ATen/native/CPUBlas.cpp:
```diff
@@ -327,6 +327,36 @@ void gemm(
+      beta,
+      bfloat_v.data(), ldc);
+  for (auto cv: bfloat_v) {
+    *(c++) = c10::convert<float>(cv);
```
This potentially truncates values between rows in `c`  if the data is strided.

mingfeima replied:
Updated, strided rows won't be updated now.

mingfeima committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"


This one is a wrapper upon `mkl_gemm_bf16bf16f32` which is used in flash attention kernel on intel 4th gen xeon.
Fallback path has also been implemented on cpublas::gemm in case `mkl_gemm_bf16bf16f32` is not available.

ref: https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-0/cblas-gemm-bf16bf16f32.html

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"


This one is a wrapper upon `mkl_gemm_bf16bf16f32` which is used in flash attention kernel on intel 4th gen xeon.
Fallback path has also been implemented on cpublas::gemm in case `mkl_gemm_bf16bf16f32` is not available.

ref: https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-0/cblas-gemm-bf16bf16f32.html

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "enable mkl_gemm_bf16bf16f32 in cpublas::gemm"


This one is a wrapper upon `mkl_gemm_bf16bf16f32` which is used in flash attention kernel on intel 4th gen xeon.
Fallback path has also been implemented on cpublas::gemm in case `mkl_gemm_bf16bf16f32` is not available.

ref: https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-0/cblas-gemm-bf16bf16f32.html

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented:
close this one as we have a replacement https://github.com/pytorch/pytorch/pull/107196
